cmd/kubeadm/app/phases/upgrade/prepull.go

```diff
@@ -183,6 +184,12 @@ func buildPrePullDaemonSet(component, image string) *apps.DaemonSet {
 					},
 					Tolerations:                   []v1.Toleration{constants.ControlPlaneToleration},
 					TerminationGracePeriodSeconds: &gracePeriodSecs,
+					// Explicitly add a PodSecurityContext to allow these Pods to run as root.
+					// This prevents restrictive PSPs from blocking the Pod creation, by automatically
```

Comment:
If a permissive PSP is present that allows the pod to be created as-is, it will be created as is (with no podSecurityContext and no explicit mustRunAsNonRoot setting)

The only time the pod would be forced to `mustRunAsNonRoot: true` by PSP is if that was the only psp available to the pod. In that case, setting this to false here would cause the pod to be completely rejected by PSP, and the pod could not be created at all. That might be what you want, since it indicates a permission problem earlier, but setting this to false will not prevent a restrictive PSP from blocking the pod creation.

Reply:
Unfortunately that is not what happens. PSPs get evaluated in order, so it makes sense to have a restricted PSP with a `MustRunAsNonRoot` rule as the first in order. In that case, with no `runAsNonRoot` field set in the SecurityContext, the pod will be forced to run as non root by the PSP. This can be seen in the following code, which specifically triggers on `sc.RunAsNonRoot() == nil` and sets its value to `true`:

https://github.com/kubernetes/kubernetes/blob/ace60e8acb2c62e006884a654936d7e30a67824b/pkg/security/podsecuritypolicy/provider.go#L167-L173

Kubelet evaluates the actual runtime user (i.e. the value specified by the `USER` command in the Docker image) and sees that the container will be run as root, causing pod creation to fail.

Reply:
> PSPs get evaluated in order, so it makes sense to have a restricted PSP with a MustRunAsNonRoot rule as the first in order. 

Non-mutating PSPs are preferred over mutating ones

Reply:
Then it seems like that preference isn't working out like it should. When I create this pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  containers:
  - name: test
    image: debian
```
on a cluster with [these PSPs](https://gist.github.com/praseodym/88ea84fef721f0da5ecf6b1f42ba77d7#file-psp-yaml) configured, the pod gets admitted to the `a-restricted` PSP with a `MustRunAsNonRoot` rule:
```
I0513 19:34:54.344576       1 admission.go:135] pod test (generate: ) in namespace kube-system validated against provider a-restricted
```
That log line is logged by a part the PSP admission plugin where mutation is allowed:

https://github.com/kubernetes/kubernetes/blob/954996e231074dc7429f7be1256a579bedd8344c/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L127-L135

And indeed, the pod never starts because was mutated to have a `runAsNonRoot: true` SC:
```
  Warning  Failed     60s (x8 over 2m22s)  kubelet, k8s-06    Error: container has runAsNonRoot and image will run as root
```

Reply:
> Then it seems like that preference isn't working out like it should. When I create this pod:
> ...
> on a cluster with these PSPs configured, the pod gets admitted to the a-restricted PSP with a MustRunAsNonRoot rule:

I see. In your case, you have multiple policies which require mutating the pod, so no policy allows the pod as is. In your cluster, you've configured the restrictive policy to take precedence over the moderate policy.

The full logic is:
* If any authorized policies allow the pod as-is, admit the pod with no changes
* If any authorized policies allow the pod with mutations, admit the pod under the first policy alphabetically

In your case, both the `a-restricted` and `c-moderate` policies would allow your example pod *if it had some changes applied* (dropped capabilities, etc). Since `a-restricted` comes first alphabetically, the restricted policy was chosen.

By explicitly setting `securityContext: {runAsNonRoot: false}`, your pod is no longer permitted at all by the `a-restricted` policy, so the `c-moderate` policy is the only one that would allow it.

If you want the moderate policy to be preferred for pods that it would allow, it would be simpler to name it so that it comes before your restrictive policy.

If the image prepull pod requires root access, then setting a `runAsUser: 0` in the pod spec is fine. Like I said, it will not allow the pod to run in a scenario where the only PSPs available required running as non root.

Reply:
TIL! Thanks Liggitt.